### ISSUE TITLE
Problems building ieee.c in Hercules using SoftFloat-3a public headers.

### ISSUE THREAD

**Issue by srorso** [hercules-390/SoftFloat-3a#5]
From Peter Coghlan in Hyperion issue [#224](https://github.com/hercules-390/hyperion/issues/224):

"Both softfloat.h and softfloat_types.h attempt to #include <stdint.h> without checking HAVE_STDINT_H

See hstdint.h for a more portable solution.

Typedefs for (u)int_fast8/16/32/64_t are still missing after that but that is easier to work around."

**Comment by srorso** [comment]
The CMake for SoftFloat-3a takes pains to ensure a stdint.h is available for the SoftFloat-3a build and for the Hercules build, to the point of creating one if it does not exist.  These headers enable Hercules compilation and include the types  noted by Peter.  See stdint.h.in in the Softfloat-3a directory.

Research continues into why this file was not processed by CMake when it is run on a system without the required stdint.h.

**Comment by PeterCoghlan** [comment]
Many thanks for looking into this for me.

I am not using CMake.  I am building Hercules on VMS using my own build procedure.

My understanding is that ``#include <stdint.h>`` should look for stdint.h in system include libraries only, not in include directories specified by the build procedure.  The latter would require ``#include "stdint.h"``.

Precedent in other parts of Hercules suggests that ``#include <stdint.h>`` should only be attempted if HAVE_STDINT_H is defined and something different should be attempted if HAVE_STDINT_H is not defined.  Perhaps there are reasons for not doing this in the case of softfloat.h and softfloat_types.h though?

**Comment by srorso** [comment]
Hi Peter:  

Apparently, #include <> also includes directories named in -I compiler command line options, and the autotools install includes a -I for the SoftFloat-3a public headers directory; this is where the ersatz stdint.h gets created if the target does not have one.  

SoftFloat-3a requires the (u)int_fast8_t (etc) headers.  If SoftFloat-3a For Hercules were to be changed to use an include structure similar to that in hstdint.h, would that address building Hercules under VMS?  

Alternatively, if you are using the install directory structure for SoftFloat-3a and have a -I for the include directory in that structure, you could copy stdint.h.in from SoftFloat-3a and rename it to stdint.h.  

It is expected that the Hercules build retrieves the SoftFloat-3a headers from the -I directory, not from the Hercules source directory, although I see I did the ieee.c include incorrectly for that ("" instead of <>).  The SoftFloat-3a headers exist in the Hercules directory to enable IDEs that incrementally compile to produce useful results.  

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
I've had a look through the source and the only place I can find hstdint.h included is in decContext.h, part of decNumber.  Rather than have two different mechanisms to address the availability or not of stdint.h on different platforms for both decNumber and Softfloat, I think it makes sense to have just one, either using the existing hstdint.h method or a new method.

Looking at the layout of hstdint.h, it suggests that Windows/MSVC might not have stdint.h, inttypes.h or unistd.h either.  I wonder if this is really the case? VMS has inttypes.h which defines the types listed in the comments at the start of hstdint.h but not the (u)int_fast*_t types.  There must be platforms platforms other than VMS which do this too as hstdint.h was created in Hercules without me having to ask for it. I guess there must be other platforms which define useful types in unistd.h too but I don't know which platforms these are.  (VMS has a unistd.h too but it doesn't contain any of this stuff.)

I wonder if it is a good idea to create something called stdint.h that is not the same as the "real" stdint.h though?  Perhaps it should be called something else?  Maybe this is how hstdint.h originally came to be?

Perhaps the existing hstdint.h could be used for both decNumber and Softfloat by adding typedefs for the required (u)int_fast*_t types in the ``#if defined(_MSVC_)`` and ``#elif defined(HAVE_INTTYPES_H)`` sections?

If it is necessary for other reasons to create a file, perhaps hstdint.h is the file to create?

**Comment by srorso** [comment]
Further research:

A basic premise of the construction of the SoftFloat-3a For Hercules (S3FH) repository is that S3FH compile without reference to any component of Hercules.  There are several reasons for this.  The only connection between S3FH and Hercules is the provision to Hercules of a statically linkable library, preferrably compiled to enable inclusion in a shared library (meaning PIC), and such public headers (two, `softfloat.h` and `softfloat-types.h`) in an interface includes directory.  

(Use of `hconsts.h` in the decNumber build will be addressed in due course.)

The original SoftFloat-3a did not guard its includes for `stdint.h` and `stdbool.h` and needs them to compile source and to provide complete interface headers.  

S3FH guards the includes for `stdint.h` and `stdbool.h` by the presence of the types defined in these headers, and if the headers do not exist, creates substitute `stdint.h` and `stdbool.h` headers in the S3FH interface includes directory.  Creation of the substitute headers enables compilation of S3FH, and inclusion of the substitute headers in the interface includes directory should allow the S3FH public headers to compile correctly within Hercules.  

Although the #include statements within S3FH, both source and public headers, use <> to include `stdint.h` and `stdbool.h`, modern (however defined) search the interface include directories (-I) _after_ the system directories if the #include'd file does is not found in the system directories.  

All of which does not seem to work in your environment.  

I agree that creating substitute headers with the same name as system headers may not be the best form, and I agree that the #include sequences in source code, both SoftFloat-3a and s3FH, might be improved.  And I am a sucker for getting Hercules to build on a new platform, even if it is an old platform.  Even if the attempt fails, the attempt improves the portabilty of the Hercules (and S3FH) code.  

So I have the following questions....

* What specifically is failing?  In particular, is it a failure of S3FH to compile or do the S3FH public headers fail to compile in ieee.c?

* Are the `[u]int.._t` types defined elsewhere in your system's include set?  In other words, if `stdint.h` is not present on your system, can the S3FH build infer that the types normally defined therein do not exist and elect to define them internally?  Likewise for the types defined in `stdbool.h`.

* Does your c compiler search interface include libraries for <>-style includes.  

* How are you addressing the absence of `stdint.h` in the build of Hercules?  Hercules requires those types.  

* Error messages from a failed compilation unit (`ieee.c` or one of the S3FH units) would be helpful, even if the error messages are quite different from those issued by a modern (however defined) c compiler.  

* Should I be able to provide updated code for your functional testing, what would you like to receive.  At the moment, updates might include a new `platform.h(.in)` and perhaps differently-named versions of `stdint.h` and `stdbool.h`.  I am thinking about this in `platform.h`.
```
#ifdef HAVE_STDINT_H
#include <stdint.h>
#else
#include "sfstdint.h"    /* or outright type definition; there are not that many  */
#endif
```
Ditto for `stdbool.h`, and something similar would be included in the `softfloat.h` public header.  Something similar because if your Hercules build already defines the `[u]int.._t` types, then the conditionals in  `softfloat.h` will need to be slightly different.  

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

The finer details of this issue have gone out of my brain-cache so I need to do some more investigation at my end.  However, I will try to answer some of your questions before I I find time to do this.

The specific issue is that compilation of ieee.c fails due to not finding include file stdint.h

My system has inttypes.h which typedefs (u)int8_t, (u)int16_t, (u)int32_t, (u)int64_t and (u)intptr_t (only).
It does not create any types with "fast" in their names.

I suspect my compiler should be able to search for files included using angle brackets in arbitrary directories - I need to look into it.  I am unsure of the meaning of "interface include libraries".  Does this just boil down to a specified directory containing include files which is not the directory containing the source being built and not the directory where the build is being performed?

On my system, the (u)int..t types required by Hercules (excluding softfloat) are provided by inttypes.h which gets included by hstdinc.h.  I think Linux / Unix builders get them there too because as far as I can see, Hercules does not seem to include stdint.h anywhere except when compiling code related to softfloat and possibly decnumber.  I am unsure where Windows builders pick up these types but looking at hstdint.h suggests that they may be defined in that file when building on Windows.

As far as I recall, I don't have any issues with stdbool.h.  It exists on  my system, the complier seems to find it and it apparantly contains whatever Hercules needs it to contain.  Do you suspect I could be having issues with stdbool.h?

**Comment by jphartmann** [comment]
Peter, I am sure you are aware of the compiler option -I (that is, capital i) and that it does not fix the problem.  Have you tried -isystem ?  I suppose the last resort is -nostdinc

**Comment by srorso** [comment]
Hi Peter:

You have provided a wonderful cache of information, maybe enough to move forward with a solution that works.  Finding one that works well (_i.e.,_ is portable and not a one-off for your environment) may take a bit longer.   

SoftFloat-3a documents a requirement for the following types:
```
bool
uint16_t
uint32_t
uint64_t
int32_t
int64_t
uint_fast8_t
uint_fast32_t
uint_fast64_t
```
Also required, but not documented, is the macro `false` (nominally from `stdbool.h`) and `uint_least8_t`, used to define an array in the function softfloat_countLeadingZeros8().  

The uint_fast(8|32|64)_t types are used to allow the c compiler to optimize to the hardware architecture.  SoftFloat-3a uses them for internal calculations and for internal communications.  The [u]int(16|32|64)_t types are used for the public API, and those are by definition 16, 32, or 64 bits long.  The `fast` versions are _at least_ 8, 32, or 64 bits long.  On a DEC Alpha, I will bet that `uint_fast8_t` would have been defined as 32 bits (longword in Alpha parlance) because there are only a couple of instructions that operate on shorter than 32-bit entities.  I found the [Alpha Architecture Handbook](https://www2.cs.arizona.edu/projects/alto/Doc/local/alphahb2.pdf) helpful on this matter, starting on p. 2-1.  

At the same time, defining `uint_fast8_t` as `uint_8` may or may not exact a performance penalty within SoftFloat-3a.  SoftFloat-3a uses `uint_fast8_t`to type a variable used to store a local copy of the rounding mode for an operation; little arithmetic is performed on that local copy.  And if you are not going to actually _do_ IEEE binary floating point in the built Hercules, then there is no penalty at all.  :-)

So the challenge becomes: In the absence of `stdint.h`, how to define the four types provided by `stdint.h` that are not defined in `inttypes.h`.  Should be doable.  

I cannot just change SoftFloat-3a For Hercules to use "" includes for its own versions of stdbool.h and/or stdint.h instead of <> includes when stdint.h is not available.   Because `inttypes.h` defines _some_ of the required types but not all, inclusion of a bootleg `stdint.h` will multiply-define the non-fast and non-least types.  

Right now I am looking at the addition of an `sfstdinc.h` header in the S3FH interface include directory (easy), source code changes to 273 compilation units (a single `bash` command will do it), some work on the CMake build (for portability), and changes to the odd half-dozen header files.  The `sfstdinc.h` header can sort the difference between using `stdint.h` or using `inttypes.h` and adding definitions for the `fast` and `least` types.  

Your issue has also led me to realize that I have an issue using the `fast` types in the public interface header softfloat.h.  I need to review that further.   

I am using the term "interface include directory" just as you suspect: a directory specific to a separately-built package that contains the public headers needed to use the package.  In UNIX-like systems, this seems to be `/usr/include` or `/usr/local/include` for installed packages.  Hercules does not presume to install S3FH when it is building S3FH, so a specific interface include directory needs to be passed to teh c compiler using -I (dash-eye).  And Windows is, of course, different.  

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
After some testing I have found that specifying the /include_directory qualifier to the compiler does indeed allow it to find files included using angle brackets in the specified directories.  This will allow me to pick up a specially created stdint.h from elsewhere as required by the new Hyperion build environment.  My apologies for not spotting this sooner.  It appears there is the makings of a solution in this if there are no other options.

Could it turn out that the "fast" types are not required in the public interface header softfloat.h? Would that mean that including inttypes.h there instead of stdint.h would suffice?  If this would be sufficient in all environments including Linux/unix, Windows and VMS, perhaps that might be the way to go for maximum portability?  However, looking at the number of int_fast.._t types in softfloat.h suggests this is probably too much to hope for and the sfstdinc.h method you outlined above is probably the way to go.  If that doesn't prove workable either, we could fall back to the creating a bootleg stdint.h file method as a last resort.

**Comment by srorso** [comment]
Ok, having done further research, I discover I have really stepped in it.  

Here's the issue: every compilation unit in the SoftFloat-3a conditions the inclusion of stdint.h as follows:
```
#if !defined(int32_t) 
#include <stdint.h>             /* C99 standard integers */ 
#endif
```
Trouble is, `int_32` is not a preprocessor macro.  It's typedef'd, and that doesn't happen until _after_ the preprocessor runs.  So the statement always evaluates true, and `stdint.h` gets included.  Six times per compilation unit.  

And while the above does not achieve the intended effect of guarding the include, it is an improvement on the original SoftFloat, which did no guarding at all.  

There are two options here:

1. Bootleg stdint.h.  The following should be enough:
```
     typedef unsigned char      uint_fast8_t;
     typedef unsigned int       uint_fast32_t;
     typedef unsigned long long uint_fast64_t;
     typedef unsigned char      uint_least8_t;
     typedef signed   char      int8_t  /* needed for Hercules */
     typedef unsigned char      int8_t  /* needed for Hercules */
     typedef          short     int16_t;
     typedef unsigned short     uint16_t;
     typedef          int       int32_t;
     typedef unsigned int       uint32_t;
     typedef          long long int64_t;
     typedef unsigned long long uint64_t;
```
If you are willing, would you test this bootleg for completeness?  Additional conditionals are needed for a general purpose portable solution that does not generate "benign redefinition of type" warnings.  

2. Take the softfloat_includes.h route, change 273 programs, refactor the 12 lines of includes at the start of each program into softfloat_includes.h, and change the remaining headers so they do not redundantly include things.  This will take longer, but I believe I will be doing this anyway.  The hardest part about this is setting up a personal fork of the Softfloat-3a for Hercules repo and building on lots of systems.  Which is to say not very hard but a bit time-consuming.

**Comment by PeterCoghlan** [comment]
Hi Steve,

This stuff is a minefield!

I tried compiling ieee.c with the above bootleg stdint.h and I ran into some problems.

int_fast16_t and int_fast32_t are still missing.

The compiler complains that typedefs for `int64_t` and `uint64_t` conflict with different typedefs for `int64_t` and `uint64_t` which are picked up from inttypes.h. (It turns out that the compiler will tolerate multiple typedefs defining the same type provided they are exactly the same as each other, issuing an informational message about this if directed to.  I wonder is this what you mean by "benign redefinition of type" warnings?)

If I ensure `HAVE_INTTYPES_H` is undefined in an effort to not have inttypes.h included, I run into these errors:
```
    #error Unable to find fixed-size data types
    #error Unable to define u_intNN_t data types
```
which are both issued by htypes.h and then I find that inttypes.h gets included anyway via bitypes.h which in turn gets included by netdb.h for example as well as others.  So, it looks like I am stuck with having inttypes.h included whether I want it or not.

I can get ieee.c to compile ok if I use the following stdint.h although only the "fast" types are strictly necessary:

```
     typedef unsigned char      uint_fast8_t;
     typedef signed   short     int_fast16_t;
     typedef unsigned int       uint_fast32_t;
     typedef signed   int       int_fast32_t;
     typedef unsigned __int64   uint_fast64_t;
     typedef signed   __int64   int_fast64_t;
     typedef unsigned char      uint_least8_t;
     typedef signed   char      int8_t;  /* needed for Hercules */
     typedef unsigned char      uint8_t;  /* needed for Hercules */
     typedef          short     int16_t;
     typedef unsigned short     uint16_t;
     typedef          int       int32_t;
     typedef unsigned int       uint32_t;
     typedef          __int64   int64_t;
     typedef unsigned __int64   uint64_t;
```

**Comment by srorso** [comment]
Hi Peter,

This is excellent information.  I missed the two int_fast.._t types; sorry about that.  We are getting much closer, though, to an ultimate solution.  

Would you now give this a try:

1) Replace softfloat_types.h in both the SoftFloat-3a For Hercules source directory and in its interface include directory with the attached.  (Be sure to get rid of the .txt suffix, needed to work around github's file type restrictions.)

2) Make sure when building SoftFloat and Hercules that HAVE_STDINT_H is _undefinded_ and HAVE_INTTYPES_H _is_ defined.  

I have been testing the updated softfloat_types.h on my systems, although all have stdint.h.  So the testing would just reveal selected typos.  The updated header also defines int32_t as a macro so the guards in the rest of softfloat work.  

And your description is exactly what I mean by "benign redefinition of types."  

[softfloat_types.h.txt](https://github.com/hercules-390/SoftFloat-3a/files/1320915/softfloat_types.h.txt)

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

I have not gotten around to trying to compile SoftFloat-3a For Hercules and I haven't created a build procedure that will enable me to compile it on my system yet.  This looks fairly straightforward to do but I need a little time to do it.  I have been concentrating on being able to compile Hercules itself first.

I tried replacing the softfloat_types.h in the Hercules source directory with your new one above and then compiling ieee.c.  The compiler still complained that it couldn't find stdint.h (included by softfloat.h) so I created an empty stdint.h to quieten it for now.  Then it complained that `bool` was being redefined and the new definition was not compatible with the old one.  I got around that by defining `HAVE_STDBOOL_H` but perhaps there should be a check to see if `bool` is already defined before redefining it in any case?

At this point, I think compilation would have succeeded except that int_fast64_t is still missing.

I am curious why the compiler doesn't notice/object to softfloat_types.h having
```
       typedef          long long int64_t;
       typedef unsigned long long uint64_t;
```
which is different from inttypes.h having
```
   typedef __int64 int64_t;
   typedef unsigned __int64 uint64_t;
```
and whether there are any consequences of the difference.

**Comment by srorso** [comment]
Hi Peter:

Give this a try to compile ieee.c.  [softfloat_types.h.txt](https://github.com/hercules-390/SoftFloat-3a/files/1324569/softfloat_types.h.txt)

To build SoftFloat-3a for Hercules, you will also need an updated platform.h, and that is being developed.

While Hercules does not require C99 headers, SoftFloat clearly does, and the changes to softfloat_types.h should enable compilation of the S3FH public headers on a  non-C99 system.   This includes your system, I think, and Windows systems prior to Visual Studio 2015.  The current committed version of S3FH addresses the issue for Windows but not UNIX-like systems.  

With respect to your question about the types in inttypes.h versus those in softfloat_types.h, I believe (if I coded it correctly) the typedefs in softfloat_types.h are skipped on systems that have inttypes.h.  

Anyway, give this a try.  

Best Regards, 
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

I can successfully compile ieee.c using the new softfloat_types.h plus an empty stdint.h

Thanks for the clarification regarding avoiding conflicts with types already defined in inttypes.h.  I appear to have been looking at how you did it but not seeing it :-)

**Comment by srorso** [comment]
Hi Peter,

Is the empty stdint.h still needed?   I think I fixed that but do not know.  It is impossible to test on a C99 system.   

As for looking and not seeing, well, if I had a nickle for each time l’ve done that....

**Comment by PeterCoghlan** [comment]
ieee.c #includes softfloat.h which in turn #includes stdint.h before #including softfloat_types.h which #defines int32_t to which guards against #including stdint.h but the #include stdint.h has already been attempted and failed because the file does not exist.

**Comment by srorso** [comment]
Hi Peter,

Please recall my prior comment about "if I had a nickle..."  :-)

Anyway, you may delete the following offending lines from softfloat.h:
```
#if !defined(false) 
#include <stdbool.h> 
#endif
#if !defined(int32_t) 
#include <stdint.h>             /* C99 standard integers */ 
#endif
```
Softfloat_types.h is not in final form; I built S3FH, tested, and discovered during testing two failures in scaled results returned for an extended precision multiply or divide with IEEE traps enabled.  Turns out that if `[u]int_fast16_t` _really is_ just 16 bits wide, the  128-bit scaled result routine returns a NaN rather than the correct scaled result. Redefining `[u]int_fast16_t` as a 32-bit makes the problem go away, but I need to investigate what I have coded incorrectly in f128_scaledResult.c.  

Apparently, on Intel at least, the fast way to do halfword arithmetic is to use a fullword.  And this would appear to be true on Alpha.  

This issue will not prevent clean compilation.

**Comment by PeterCoghlan** [comment]
I deleted the section from softfloat.h and recompiled ieee.c.  Then I realised that I forgot to delete the empty stdint.h so I did that and recompiled again.  It works :-)

I compiled the rest of hercules as well and didn't run into any new compile time issues.  There usually are some when I have been away from it a while so I am pleased.  No run time issues that I have noticed either.

**Comment by srorso** [comment]
Hi Peter,

That is excellent news.  And you mentioned no runtime issues...does that mean you built SoftFloat-3a For Hercules?  And if not, how did you keep the absence of S3FH from being a problem?  Last question: is there a REXX implementation on your system that would allow for `runtest` or the equivalent of `make test`.

Best Regards, 
Steve Orso

**Comment by jphartmann** [comment]
`make check` is the UNIX way.  Have you changed it, Steve.

**Comment by PeterCoghlan** [comment]
I didn't build SoftFloat-3a For Hercules yet.  I decided to wait for the new platform.h. I just link everything I have built and leave out what I don't have for now.  The result is an executable with 70 undefined symbols, 66 of which are part of S3FH.  (The other 4 are CTC networking devices which I don't have implemented.)  The executable runs just fine as long as I don't try to execute any of the instructions implemented by S3FH and don't specify one of the unimplemented networking devices in my Hercules config.  (I have a friend who writes code for the more popular platforms and who exclaims in horror when I relate stuff like this.)

I believe there is a Regina REXX for VMS and there may be other implementations but I don't have any set up at the moment.  I will see if I can get one running.

**Comment by srorso** [comment]
Hi Peter,

Please test a compilation or two of S3FH using the following three headers:

[softfloat_stdTypes.h.txt](https://github.com/hercules-390/SoftFloat-3a/files/1329403/softfloat_stdTypes.h.txt)
[softfloat_types.h.txt](https://github.com/hercules-390/SoftFloat-3a/files/1329402/softfloat_types.h.txt)
[platform.h.in.txt](https://github.com/hercules-390/SoftFloat-3a/files/1329405/platform.h.in.txt)

Remove the .txt extension from each of the three.  In addition, remove the .in extension from platform.h and edit the #cmakedefine statements into #define statements that reflect the headers available on your system.   I suspect you will need this:
```
#define HAVE_STDBOOL_H 1
/*  #undef HAVE_STDINT_H   */
#define HAVE_INTTYPES_H 1
```
`softfloat_stdtypes.h` is faking out the `#include <stdint.h> in each compilation unit by `#define int32_t`, so you should be able to compile without an empty `stdint.h`.  The correction I plan to commit to S3FH will remove the unneeded includes from each source file.  The fakeout just gets us a test of the planned approach. 

Best Regards,
Steve Orso

p.s.  I replied to John's note above via e-mail, and I do not see it here.  So...

CMake does not create a check target from its add_test commands, it creates a test target.  I could probably add code to the build to create a make check that does the same thing as make test.

**Comment by PeterCoghlan** [comment]
Hi Steve,

I can compile the vast majority of the modules of S3FH using the new include files you provided above and I have not run into any stdint.h or types related issues.

It seems I may need a specialize.h as many modules include this file and fail to compile if it is not present.  I see specialize.h files in the 8086 and 8086-sse directories which might be suitable starting points but I am unsure what changes I might need to make to them or where I should put the resulting file.

It looks like I may also need to compile some of the .c files that are in the 8086 or 8086-sse directories as I seem to have some unresolved symbols when I try to link Hercules without them (unless I have messed up somewhere which is also possible).

**Comment by srorso** [comment]
Hi Peter,

Give the 8086-SSE directory a shot.  You will need specialize.h and any file that _does not_ contain “extF80” or “extF80M” in it.  

The 8086 directory has routines for older Intel processors and was not mod’d for Hercules. 

The extF80[M] routines are for Intel x86 Floating point and are not needed by Hercules. 

And code in the hercsource directory replaces same-named code in source. 

Good luck!

Best Regards, 
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

I have identified 74 files in the source directory, 6 in the 8086-SSE directory and 25 in the hercsource directory which when compiled make the linker happy.  I had small issues compiling a handful of the other files in the source directory before finding that I didn't need them.  The only thing even approaching an issue I had with compiling the files I did need was extraneous semicolons on:
```
#define init_detectTininess softfloat_tininess_beforeRounding;
```
and
```
#define init_detectTininess softfloat_tininess_afterRounding;
```
variants of which crop up in specialize.h and softfloat-specialize.h in the 8086-SSE and 8086 directories.

Now that everything is compiled successfully, I've tried running some tests but I've run into some problems there which I will look into tomorrow.

**Comment by srorso** [comment]
Hi Peter, 

I get 6 in 8086-SSE, 25 from hercsource, and 119 from source, but that does not take into account  inlined functions that replace compilation units and the fact that I am likely compiling things that are not required.  Although they don't link in to libherc.so, so no downstream impact.  Can you post or PM the 75 from source?

Thanks for pointing out the semicolons; my mistake, and yours is the first compiler to complain.  It shall be corrected.  

I am thrilled to see that you are making progress.  

Best Regards,
Steve Orso

**Comment by srorso** [comment]
Hi Peter,

I was able to build S3FH with just the 74 files in the `source` directory, build Hercules with that S3FH library, and run a clean set of tests.  I shall consolidate all of the activity into a commit of my temporary fork of S3FH--that will take a few days--and I will ask you to clone that fork for an end-to-end test.  Most of the changes will not affect you, but as you have the only C90 system I know about, I welcome a chance to make sure I did not break anything for you.  

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

Many thanks for sorting this out for me.  I greatly appreciate being able to continue to grab the latest Hyperion and quickly build it on my system with only minimal and easily automated "porting" required. I hope that the changes made will also make it easier for others on less mainstream platforms to keep up with current Hyperion.

Let me know when you are ready for me to test further.

The compiler I am using is supposedly configurable for ANSI89, ISO C94, C99 and a variety of other modes but clearly the C99 implementation is incomplete in respect of at least some header files.

Regards,
Peter Coghlan.

**Comment by srorso** [comment]
Hi Peter,

I have committed the "final" version of the updates for SoftFloat-3a For Hercules to my temporary fork of the S3FH repository; find that temporary fork here:

  [https://github.com/srorso/SoftFloat-3a](https://github.com/srorso/SoftFloat-3a)

It has been tested through make check/make test on a number of UNIX-like systems and on Windows 10 using Visual Studio 2017.  But all of those systems are nominally C99, so your system is the real test.  

To all: upon feedback from Peter, I will commit these changes to the proper S3FH repository along with one co-requisite change to the CMake build for Hercules (the CMake build does not work on pre-C99 systems either).

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

I've built SoftFloat-3a For Hercules from your temporary fork above and it works great.  The only thing I came across was that the extraneous semicolons are still there resulting in an informational message when compiling softfloat_state.c

I've run tests bfp-001 to bfp-022 and all fully passed.

Regards,
Peter Coghlan.

**Comment by srorso** [comment]
Peter,

Again, excellent news...except for that semicolon...I have updated my temp fork, and just to make sure I have not messed it up, could I trouble you to give it another try?

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

I've built from your temporary fork again and it works.  No errors.  No warnings.  No informational messages.  All good.

I've also rerun tests 1 to 22 and they all still work too.

Many thanks.

Regards,
Peter Coghlan.

**Comment by srorso** [comment]
See also commit [https://github.com/hercules-390/hyperion/commit/d660499da8f498ba53e6082033ae0ec3b540f812](https://github.com/hercules-390/hyperion/commit/d660499da8f498ba53e6082033ae0ec3b540f812)

**Comment by srorso** [comment]
Hi Peter,

With your confirmation, I would like to close this as having been addressed.  Should you experience further difficulties with this matter, please feel free to re-open.  

Best Regards,
Steve Orso

**Comment by PeterCoghlan** [comment]
Hi Steve,

Many thanks for all the work you have done on this issue.

I can confirm that I have no difficulties compiling Softfloat 3a for Hercules now.

However, a day or so after the Oct 2 commit for Hyperion above, I went back to trying to compile ieee.c again and it still failed due to not being able to find stdint.h.  Perhaps this is yet to be addressed under Hyperion issue #224 though?  If there have been more recent changes which might have affected this, I can grab the latest Hyperion and try again.
